fvm/errors/errors.go

```diff
@@ -85,3 +86,44 @@ func HandleRuntimeError(err error) error {
 	// All other errors are non-fatal Cadence errors.
 	return NewCadenceRuntimeError(&runErr)
 }
+
+// FirstOrFailure returns:
+// - nil, if both errors are nil
+// - the non-nil error, if only one is nil
+// - the first error, if the second is not a failure, or both are a failure
+// - the second error otherwise
+// In case both err1 and err2 are non-nil, the one that is hidden gets logged.
+func FirstOrFailure(logger zerolog.Logger, err1 error, err2 error) error {
+	// if either is nil, return the other
+	// this is not hiding any errors
+	if (err1 == nil) != (err2 == nil) {
+		if err1 != nil {
+			return err1
+		}
+		return err2
+	}
+
+	// both are nil return nil
+	if err1 == nil {
+		return nil
+	}
+
+	// if first error is a failure or second is not a failure, return first error, hide the second
+	var failure Failure
+	if As(err1, &failure) || !As(err2, &failure) {
+		logger.
```

Comment:
Instead of just logging the masked error, we should return both errors.

looks like we already use "github.com/hashicorp/go-multierror" elsewhere in the codebase

Reply:
Not sure how that would work with error codes and unwrapping... let me check.

Reply:
I'm still not entirely sure how returning a multi-error from this location would complicate downstream error handling.
This error eventually reaches the user and is added as the transaction result, which means users would have to know this could happen and handle it appropriately. 

From a users perspective, it might not be easy to know what to do if you get multiple errors...

any thoughts here: @zhangchiqing @ramtinms @m4ksio

Reply:
The user should handle the error by error code, rather than by the error message.  When I said use multi-err, I meant, keep both error messages, but provide the scarier of the two error codes.

The user would already need to handle both cases since you can return either one today

Reply:
Error code is currently not a separate field on the transaction result. So users have to check the error string for the code.

Reply:
that's unfortunate.  we still return both error messages, but strip the less useful error code?